### PR TITLE
Defer the OverlayEntry listenable disposal until its widget is unmounted

### LINKED ISSUE #102718: Calling dispose() on OverlayEntry throws an assertion error
An assertion error is thrown when `dispose()` is used after `remove()`.

```dart
_overlayEntry?.remove();
_overlayEntry?.dispose();
```

I think this is because `_markDirty()` is called from `remove()` in the next frame when `dispose()` has already been executed.

https://github.com/flutter/flutter/blob/d873157287ec4bdff8195581dd8167e557acc0f5/packages/flutter/lib/src/widgets/overlay.dart#L154-L158

The error is avoided if `dispose()` is also delayed to the next frame manually like below, whether it is good or not.

```dart
_overlayEntry?.remove();
SchedulerBinding.instance.addPostFrameCallback((_) => _overlayEntry?.dispose());
```

I guess that Flutter devs probably think they should use `dispose()` when an OverlayEntry becomes no longer necessary as they usually do when they finish using a `ChangeNotifier`. The unexpected error has been confusing me. Should `dispose()` not used? (Is it enough to use only `remove()`?) If so, I think it should be documented.

## Steps to Reproduce

1. Execute `flutter run` on the code sample below.
2. Press the FAB to show an overlay.
3. Press the FAB again to hide the overlay.

**Expected results:** 

No error.

**Actual results:** 

The last step above throws an error, showing `A OverlayEntry was used after being disposed.`

<details>
<summary>Full output</summary>

```
======== Exception caught by widgets library =======================================================
The following assertion was thrown while finalizing the widget tree:
A OverlayEntry was used after being disposed.

Once you have called dispose() on a OverlayEntry, it can no longer be used.
When the exception was thrown, this was the stack: 
#0      ChangeNotifier._debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:114:9)
#1      ChangeNotifier._debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:120:6)
#2      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:288:12)
#3      OverlayEntry._updateMounted (package:flutter/src/widgets/overlay.dart:130:5)
#4      _OverlayEntryWidgetState.dispose (package:flutter/src/widgets/overlay.dart:200:18)
#5      StatefulElement.unmount (package:flutter/src/widgets/framework.dart:4983:11)
#6      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1926:13)
#7      ListIterable.forEach (dart:_internal/iterable.dart:39:13)
#8      _InactiveElements._unmountAll (package:flutter/src/widgets/framework.dart:1935:25)
#9      BuildOwner.lockState (package:flutter/src/widgets/framework.dart:2519:15)
#10     BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:2932:7)
#11     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:884:19)
#12     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:363:5)
#13     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
#14     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1081:9)
#15     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:995:5)
#19     _invoke (dart:ui/hooks.dart:151:10)
#20     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:308:5)
#21     _drawFrame (dart:ui/hooks.dart:115:31)
(elided 3 frames from dart:async)
====================================================================================================
```
</details>

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const App());

class App extends StatelessWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  OverlayEntry? _overlayEntry;

  @override
  void dispose() {
    _hideOverlay();
    super.dispose();
  }

  void _hideOverlay() {
    _overlayEntry?.remove();
    _overlayEntry?.dispose();
    setState(() => _overlayEntry = null);
  }

  void _showOverlay() {
    final entry = OverlayEntry(
      builder: (_) => const Center(child: Text('Overlay')),
    );
    Overlay.of(context)!.insert(entry);
    setState(() => _overlayEntry = entry);
  }

  @override
  Widget build(BuildContext context) {
    final isShown = _overlayEntry != null;

    return Scaffold(
      floatingActionButton: isShown
          ? FloatingActionButton(
              child: const Icon(Icons.remove),
              onPressed: _hideOverlay,
            )
          : FloatingActionButton(
              child: const Icon(Icons.add),
              onPressed: _showOverlay,
            ),
    );
  }
}
```
</details>

<details>
  <summary>Logs</summary>

```powershell
PS D:\bug> flutter analyze
Analyzing bug...                                                        
No issues found! (ran in 1.4s)
```

```powershell
PS D:\bug> flutter doctor   
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 2.10.5, on Microsoft Windows [Version 10.0.22000.613], locale ja-JP)
[√] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[X] Chrome - develop for the web (Cannot find Chrome executable at .\Google\Chrome\Application\chrome.exe)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.1.3)
[!] Android Studio (not installed)
[√] IntelliJ IDEA Ultimate Edition (version 2022.1)
[√] VS Code (version 1.66.2)
[√] Connected device (2 available)
[√] HTTP Host Availability

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
There're customers using `OverlayEntry.addListener` (and leaking listeners unfortunately).
Alternatively we can let `_OverlayEntryWidgetState` own the value notifier (so it doesn't leak since it will always be disposed by the state object). Old listeners won't be notified when the entry is reused to build new widgets but our internal customer and route entry are only interested in the mounted state of the current widget.
Fixes https://github.com/flutter/flutter/issues/102718

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 1]
```diff
@@ -120,14 +125,21 @@ class OverlayEntry extends ChangeNotifier {
...
-    if (value == _mounted) {
-      return;
-    }
-    _mounted = value;
-    notifyListeners();
+  bool get mounted => _overlayStateMounted.value;
+
+  /// Whether the `_OverlayState`s built using this [OverlayEntry] is currently
+  /// mounted.
+  final ValueNotifier<bool> _overlayStateMounted = ValueNotifier<bool>(false);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:132]: With this change, we don't need to extend ChangeNotifier anymore, right? There's nothing from it that this class uses, AFIKT. We could just implement the Listenable interface.

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 2]
```diff
@@ -164,9 +177,40 @@ class OverlayEntry extends ChangeNotifier {
   ///
   /// You need to call this function if the output of [builder] has changed.
   void markNeedsBuild() {
+    assert(!_disposedByOwner);
     _key.currentState?._markNeedsBuild();
   }
 
+  void _didUnmount() {
+    if (_disposedByOwner && !mounted) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:185]: from the method name (_didUnmount) it should follow that mounted must always be false here? Should that `!mounted` just be in an assert?

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 3]
```diff
@@ -164,9 +177,40 @@ class OverlayEntry extends ChangeNotifier {
...
+      super.dispose();
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:186]: No functionality of the superclass appears to be used in this implementation, so there is really no reason to call its dispose method... Technically, you should dispose the _overlayStateMounted, though.

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 4]
```diff
@@ -164,9 +177,40 @@ class OverlayEntry extends ChangeNotifier {
...
+  /// tree.
+  ///
+  /// This method should only be called by the object's owner.
+  @override
+  void dispose() {
+    assert(!_disposedByOwner);
+    assert(_overlay == null, 'An OverlayEntry must first be removed from the Overlay before dispose is called.');
+    _disposedByOwner = true;
+    if (!mounted) {
+      super.dispose();
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/overlay.dart:210]: same

File: packages/flutter/lib/src/widgets/overlay.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/overlay.dart:186]: (see also the other comment about removing the super class altogether)

File: packages/flutter/lib/src/widgets/overlay.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/overlay.dart:132]: Yeah only the docs but I think I overrode the docs for `dispose` anyways.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**